File: goldenflow/cli/schedule.py
"""Schedule periodic transforms."""
from __future__ import annotations

import time
from datetime import datetime
from pathlib import Path

from rich.console import Console

console = Console()


def _parse_cron_interval(cron_expr: str) -> float:
    """Simple cron-like interval parser. Supports: '5m', '1h', '30s', '1d'."""
    expr = cron_expr.strip().lower()
    multipliers = {"s": 1, "m": 60, "h": 3600, "d": 86400}
    if expr[-1] in multipliers:
        try:
            return float(expr[:-1]) * multipliers[expr[-1]]
        except ValueError:
            pass
    raise ValueError(f"Invalid interval: {cron_expr}. Use format like '5m', '1h', '30s'")


def run_schedule(
    path: Path,
    interval: str = "1h",
    config_path: Path | None = None,
    output_dir: Path | None = None,
) -> None:
    """Run transforms on a schedule."""
    from goldenflow.config.loader import load_config
    from goldenflow.config.schema import GoldenFlowConfig
    from goldenflow.engine.transformer import TransformEngine

    cfg = load_config(config_path) if config_path else GoldenFlowConfig()
    engine = TransformEngine(config=cfg)
    seconds = _parse_cron_interval(interval)
    out = output_dir or path.parent

    console.print(f"[bold]Scheduled:[/bold] transform {path} every {interval}")
    console.print("[dim]Press Ctrl+C to stop[/dim]\n")

    run_count = 0
    try:
        while True:
            run_count += 1
            ts = datetime.now().strftime("%H:%M:%S")
            console.print(f"[cyan]Run #{run_count}[/cyan] at {ts}")
            try:
                result = engine.transform_file(path, output_dir=out)
                console.print(
                    f"  [green]Done:[/green] {len(result.manifest.records)} transforms, "
                    f"{len(result.manifest.errors)} errors"
                )
            except Exception as e:
                console.print(f"  [red]Error:[/red] {e}")
            time.sleep(seconds)
    except KeyboardInterrupt:
        console.print(f"\n[dim]Stopped after {run_count} runs.[/dim]")

File: packages/goldenflow-js/src/node/schedule.ts
import { readFile, writeFile } from "./connectors/file.js";
import { TransformEngine } from "../core/engine/transformer.js";
import { loadConfigFromString } from "../core/config/loader.js";
import { makeConfig } from "../core/types.js";
import { readFileSync } from "node:fs";
import { basename, extname, dirname, join } from "node:path";

function parseInterval(expr: string): number {
  const s = expr.trim().toLowerCase();
  const multipliers: Record<string, number> = { s: 1, m: 60, h: 3600, d: 86400 };
  const unit = s.slice(-1);
  const mult = multipliers[unit];
  if (!mult) throw new Error(`Invalid interval: ${expr}. Use format like '5m', '1h', '30s'`);
  const num = parseFloat(s.slice(0, -1));
  if (isNaN(num)) throw new Error(`Invalid interval: ${expr}`);
  return num * mult;
}

export function runSchedule(
  filePath: string,
  options: { interval?: string; configPath?: string; outputDir?: string } = {},
): void {
  const intervalStr = options.interval ?? "1h";
  const seconds = parseInterval(intervalStr);
  const cfg = options.configPath
    ? loadConfigFromString(readFileSync(options.configPath, "utf-8"))
    : makeConfig();
  const engine = new TransformEngine(cfg);
  const outDir = options.outputDir ?? dirname(filePath);

  console.log(`Scheduled: transform ${filePath} every ${intervalStr}`);
  console.log("Press Ctrl+C to stop\n");

  let runCount = 0;
  const run = () => {
    runCount++;
    const ts = new Date().toTimeString().slice(0, 8);
    console.log(`Run #${runCount} at ${ts}`);
    try {
      const rows = readFile(filePath);
      const result = engine.transformDf(rows, filePath);
      const ext = extname(filePath);
      const stem = basename(filePath, ext);
      writeFile(result.rows as Record<string, unknown>[], join(outDir, `${stem}_transformed${ext}`));
      console.log(`  Done: ${result.manifest.records.length} transforms, ${result.manifest.errors.length} errors`);
    } catch (e) {
      console.error(`  Error: ${e instanceof Error ? e.message : String(e)}`);
    }
  };

  run();
  const timer = setInterval(run, seconds * 1000);
  process.on("SIGINT", () => { clearInterval(timer); console.log(`\nStopped after ${runCount} runs.`); process.exit(0); });
}
